File: app/src/main/java/io/tripmate/ui/TripsAdapter.java
package io.tripmate.ui;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.target.Target;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.peanutsdk.recyclerview.Divided;
import io.peanutsdk.widget.CircularImageView;
import io.tripmate.R;
import io.tripmate.data.Bus;
import io.tripmate.data.Trip;
import io.tripmate.util.GlideApp;
import io.tripmate.util.LoginStateListener;
import io.tripmate.util.SearchableItem;

import static com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade;


public class TripsAdapter extends RecyclerView.Adapter<TripsAdapter.TripsViewHolder> implements
		LoginStateListener {
	
	private final Activity context;
	private final List<Trip> items;
	private final LayoutInflater inflater;
	
	public TripsAdapter(Activity context) {
		this.context = context;
		this.items = new ArrayList<>(0);
		this.inflater = LayoutInflater.from(context);
		setHasStableIds(true);
	}
	
	
	@NonNull
	@Override
	public TripsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		return new TripsViewHolder(inflater.inflate(R.layout.trip_item, parent, false));
	}
	
	@Override
	public void onBindViewHolder(@NonNull TripsViewHolder holder, int position) {
		//Get trip item per position
		Trip trip = items.get(position);
		
		//Set destination and origin
		holder.details.setText(String.format("%s - %s", trip.getOrigin(), trip.getDestination()));
		
		//Get bus details
		Bus bus = trip.getBus();
		if (bus == null) {
			//Default details when fields are null
			holder.number.setText("N/A");
			holder.name.setText("N/A");
		} else {
			//Get bus number plate
			holder.number.setText(bus.getNumber());
			
			//Get bus type
			holder.name.setText(bus.getType());
			
			//Load bus image
			GlideApp.with(context)
					.load(bus.getImage())
					.placeholder(R.drawable.avatar_placeholder)
					.error(R.drawable.avatar_placeholder)
					.fallback(R.drawable.avatar_placeholder)
					.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
					.transition(withCrossFade())
					.override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
					.into(holder.image);
		}
		
		//Set price
		holder.price.setText(String.format("GHC %s", String.valueOf(trip.getPrice())));
		
		//Set trip duration
		holder.duration.setText(String.valueOf(trip.getDuration()));
		
		//Click action for trip details
		holder.itemView.setOnClickListener(v -> {
			Intent intent = new Intent(context, TripDetailsActivity.class);
			intent.putExtra(TripDetailsActivity.TRIP_DATA, trip);
			ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(context,
					new Pair<View, String>(holder.image, "image"));
			context.startActivity(intent, options.toBundle());
		});
	}
	
	@Override
	public int getItemCount() {
		return dataItemCount();
	}
	
	@Override
	public long getItemId(int position) {
		return getItem(position).hashCode();
	}
	
	int dataItemCount() {
		int size = items.size();
		if (items.isEmpty()) return 0;
		return size;
	}
	
	@Override
	public void onLogin() {
	
	}
	
	@Override
	public void onLogout() {
	
	}
	
	
	/**
	 * Add new content to the existing list
	 * Avoid duplications by running an inner loop to check the existence of the new item in the
	 * list of existing ones. If it exists already, it will be ignored else it will be added and
	 * the dataset updated respectively
	 */
	public void addAndResort(List<? extends SearchableItem> newItems) {
		if (newItems.isEmpty()) return;
		int size = items.size();
		boolean add = true;
		for (SearchableItem item : newItems) {
			for (int i = 0; i < size; i++) {
				if (Objects.equals(item.getKey(), items.get(i).getKey())) {
					add = false;
				}
			}
			
			if (add) {
				addItem(item);
				notifyItemRangeChanged(0, newItems.size());
			}
		}
	}
	
	//Add item if it is an instance of Trip
	private void addItem(SearchableItem item) {
		if (item instanceof Trip) {
			items.add((Trip) item);
		}
	}
	
	private SearchableItem getItem(int position) {
		return items.get(position);
	}
	
	/**
	 * Clear list
	 */
	public void clear() {
		if (items.isEmpty()) return;
		items.clear();
	}
	
	public static class TripsViewHolder extends RecyclerView.ViewHolder implements Divided {
		@BindView(R.id.bus_name)
		TextView name;
		@BindView(R.id.trip_details)
		TextView details;
		@BindView(R.id.bus_number)
		TextView number;
		@BindView(R.id.trip_time_details)
		TextView duration;
		@BindView(R.id.trip_price)
		TextView price;
		@BindView(R.id.bus_photo)
		CircularImageView image;
		
		
		public TripsViewHolder(View itemView) {
			super(itemView);
			ButterKnife.bind(this, itemView);
		}
	}
}

File: app/src/main/java/io/tripmate/ui/TripsAdapter.kt
package io.tripmate.ui

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade
import com.bumptech.glide.request.target.Target
import io.peanutsdk.recyclerview.Divided
import io.peanutsdk.util.bindView
import io.peanutsdk.widget.CircularImageView
import io.tripmate.R
import io.tripmate.data.Trip
import io.tripmate.util.GlideApp
import io.tripmate.util.LoginStateListener
import io.tripmate.util.SearchableItem
import java.util.*


/**
 * Trips data adapter
 */
class TripsAdapter(private val context: Activity) : RecyclerView.Adapter<TripsAdapter.TripsViewHolder>(), LoginStateListener {
    private val items: MutableList<Trip>
    private val inflater: LayoutInflater

    init {
        this.items = ArrayList(0)
        this.inflater = LayoutInflater.from(context)
        setHasStableIds(true)
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TripsViewHolder {
        return TripsViewHolder(inflater.inflate(R.layout.trip_item, parent, false))
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: TripsViewHolder, position: Int) {
        //Get trip item per position
        val trip = items[position]

        //Set destination and origin
        holder.details.text = trip.origin + " - " + trip.destination

        //Get bus details
        val bus = trip.bus
        if (bus == null) {
            //Default details when fields are null
            holder.number.text = "N/A"
            holder.name.text = "N/A"
        } else {
            //Get bus number plate
            holder.number.text = bus.number

            //Get bus type
            holder.name.text = bus.type

            //Load bus image
            GlideApp.with(context)
                    .load(bus.image)
                    .placeholder(R.drawable.avatar_placeholder)
                    .error(R.drawable.avatar_placeholder)
                    .fallback(R.drawable.avatar_placeholder)
                    .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                    .transition(withCrossFade())
                    .override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
                    .into(holder.image)
        }

        //Set price
        holder.price.text = String.format("GHC %s", trip.price.toString())

        //Set trip duration
        holder.duration.text = trip.duration.toString()

        //Click action for trip details
        holder.itemView.setOnClickListener { v ->
            val intent = Intent(context, TripDetailsActivity::class.java)
            intent.putExtra(TripDetailsActivity.TRIP_DATA, trip)
            val options = ActivityOptions.makeSceneTransitionAnimation(context,
                    Pair(holder.image, "image"))
            context.startActivity(intent, options.toBundle())
        }
    }

    override fun getItemCount(): Int {
        return dataItemCount()
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).hashCode().toLong()
    }

    internal fun dataItemCount(): Int {
        val size = items.size
        return if (items.isEmpty()) 0 else size
    }

    override fun onLogin() {

    }

    override fun onLogout() {

    }


    /**
     * Add new content to the existing list
     * Avoid duplications by running an inner loop to check the existence of the new item in the
     * list of existing ones. If it exists already, it will be ignored else it will be added and
     * the dataset updated respectively
     */
    fun addAndResort(newItems: List<SearchableItem>) {
        if (newItems.isEmpty()) return
        val size = items.size
        var add = true
        for (item in newItems) {
            for (i in 0 until size) {
                if (item.key == items[i].key) {
                    add = false
                }
            }

            if (add) {
                addItem(item)
                notifyItemRangeChanged(0, newItems.size)
            }
        }
    }

    //Add item if it is an instance of Trip
    private fun addItem(item: SearchableItem) {
        if (item is Trip) {
            items.add(item)
        }
    }

    private fun getItem(position: Int): SearchableItem {
        return items[position]
    }

    /**
     * Clear list
     */
    fun clear() {
        if (items.isEmpty()) return
        items.clear()
    }

    class TripsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), Divided {
        val name: TextView by bindView(R.id.bus_name)
        val details: TextView by bindView(R.id.trip_details)
        val number: TextView by bindView(R.id.bus_number)
        val duration: TextView by bindView(R.id.trip_time_details)
        val price: TextView by bindView(R.id.trip_price)
        val image: CircularImageView by bindView(R.id.bus_photo)
    }
}
